### PR TITLE
feat: provide `TableDefinition` functions for Iceberg tables

### LINKED ISSUE #5868: iceberg TableDefinition from table identifier
It should be possible to get a TableDefinition (and potentially a meta_table representation) from a given table identifier without needing to read the table first.

### PR DESCRIPTION
**Pull request by lbooker42**
Will close #5868 when merged (but needs reconciled with #5880)

### CODE REVIEW

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java [hunk 1]
```diff
@@ -352,6 +353,163 @@ public Table listSnapshotsAsTable(@NotNull final String tableIdentifier) {
...
+
+    /**
+     * Return {@link TableDefinition table definition} for a given Iceberg table, with optional instructions for
+     * customizations while reading.
+     *
+     * @param tableIdentifier The identifier of the table to load
+     * @param instructions The instructions for customizations while reading
+     * @return The table definition
+     */
+    public TableDefinition getTableDefinition(
```
Comment (lbooker42) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java:419]: Only the (String tableIdentifier, final long snapshotId) are implemented. The overloads for TableIdentifier, Snapshot are not implemented.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java [hunk 2]
```diff
@@ -352,6 +353,163 @@ public Table listSnapshotsAsTable(@NotNull final String tableIdentifier) {
...
+            @NotNull final org.apache.iceberg.Table table,
+            @NotNull final Schema schema,
+            @NotNull final IcebergInstructions instructions) {
+
+        final Set<String> takenNames = new HashSet<>();
+
+        // Map all the column names in the schema to their legalized names.
+        final Map<String, String> legalizedColumnRenames = new HashMap<>();
+
+        // Validate user-supplied names meet legalization instructions
```
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java:383]:
```suggestion
        // Validate user-supplied names meet legalization rules
```

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java [hunk 3]
```diff
@@ -352,6 +353,163 @@ public Table listSnapshotsAsTable(@NotNull final String tableIdentifier) {
...
+
+    /**
+     * Return {@link Table table} containing the {@link TableDefinition definition} of a given Iceberg table, with
+     * optional instructions for customizations while reading.
+     *
+     * @param tableIdentifier The identifier of the table to load
+     * @param instructions The instructions for customizations while reading
+     * @return The table definition as a Deephaven table
+     */
+    @SuppressWarnings("unused")
```
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java:461]: Maybe we want `@ScriptAPI` annotations for this stuff?

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java (hunk 1 above)
Comment (rcaudy) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java:419]: We want the remainder, right? @devinrsmith @lbooker42

Comment (lbooker42) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java:419]: Yes, will add them. needing to support every variant of these arguments does dispose me toward an overhaul of the API.

File: extensions/iceberg/s3/src/test/java/io/deephaven/iceberg/util/IcebergToolsTest.java [hunk 4]
```diff
@@ -715,4 +715,143 @@ public void testOpenAllTypesTable() throws ExecutionException, InterruptedExcept
         final io.deephaven.engine.table.Table table = adapter.readTable(tableId, instructions);
         Assert.eq(table.size(), "table.size()", 10, "10 rows in the table");
     }
+
+    @Test
+    public void testTableDefinition() throws ExecutionException, InterruptedException, TimeoutException {
+        uploadParquetFiles(new File(IcebergToolsTest.class.getResource("/warehouse/sales/sales_multi").getPath()),
+                warehousePath);
```
Comment (devinrsmith) [inline extensions/iceberg/s3/src/test/java/io/deephaven/iceberg/util/IcebergToolsTest.java:722]: Do we need to upload parquet files to get the definition? I think it's all involved in the metadata (which we aren't uploading as part of this).

File: extensions/iceberg/s3/src/test/java/io/deephaven/iceberg/util/IcebergToolsTest.java [hunk 5]
```diff
@@ -715,4 +715,143 @@ public void testOpenAllTypesTable() throws ExecutionException, InterruptedExcept
...
+
+        // Use string and current snapshot
+        TableDefinition tableDef = adapter.getTableDefinition("sales.sales_multi", null);
+        Assert.eq(tableDef.getColumnNames().size(), "tableDef.getColumnNames().size()", 5, "5 columns in the table");
+
+        tableDef.checkHasColumn("Region", String.class);
+        tableDef.checkHasColumn("Item_Type", String.class);
+        tableDef.checkHasColumn("Units_Sold", int.class);
+        tableDef.checkHasColumn("Unit_Price", double.class);
+        tableDef.checkHasColumn("Order_Date", Instant.class);
```
Comment (devinrsmith) [inline extensions/iceberg/s3/src/test/java/io/deephaven/iceberg/util/IcebergToolsTest.java:738]: I think we should prefer to model this as

```java
TableDefinition expected = ...;
TableDefinition actual = adapter.getTableDefinition(...);
assertEquals(expected, actual);
```

or something like that

**Comment by devinrsmith** [comment]
Completed via #5891

**Pull request closed without merging**